gp_emulator/gpu/predict.cu

```diff
@@ -0,0 +1,146 @@
+/*********************************************//** 
+ * CUDA implementation of derived from GaussianProcess
+ * in python code. 
+ * 
+ * Sinan Shi (UCL) 
+ *********************************************/
+#include "gpu_predict.h"
+#include <stdlib.h>
+#define debug
+/*********************************************//**
+ * predict function:
+ * CUDA version of the corresponding python code
+ * GaussianPredict::predict
+ *********************************************/
+extern "C"{
+void predict(const real *c_theta_exp, const real *c_inputs,const real *c_invQt,const real *c_invQ, const real *c_testing,  
+        real *c_mu, real *c_var, real *c_deriv,const int N,const int M, const int  D, const int theta_size)
+{
+    int i;
+//    cublasStatus_t stat;
+    cublasHandle_t handle;
+    
+    cublasCreate(&handle);
+
+    real *c_theta_exp_sqrt;
+    c_theta_exp_sqrt = (real *)malloc( sizeof(real) * theta_size );
+    for( i=0; i < theta_size; i++ )
+    {
+        c_theta_exp_sqrt[i] = sqrt( c_theta_exp[i] );
+    }
+
+
+    //define device vector and matrices
+    real *d_inputs, *d_theta_exp, *d_theta_exp_sqrt, *d_invQt, *d_invQ, *d_testing;
+
+    //allocate and copy vector on device 
+    cudaMalloc( (void **)&d_theta_exp, sizeof(real) * theta_size );
+    cudaMalloc( (void **)&d_theta_exp_sqrt, sizeof(real) * theta_size );
+    cudaMalloc( (void **)&d_invQt, sizeof(real) * M);
+    cublasCheckErrors(cublasSetVector( theta_size, sizeof(real), c_theta_exp_sqrt, 1, d_theta_exp_sqrt, 1 ));
+    cublasCheckErrors(cublasSetVector( theta_size, sizeof(real), c_theta_exp, 1, d_theta_exp, 1 ));
+    cublasCheckErrors(cublasSetVector( M, sizeof(real), c_invQt, 1, d_invQt, 1));
+
+    //allocate and copy matrix on device
+    cudaMalloc( (void **)&d_inputs, sizeof(real) * M * D );
+    cudaMalloc( (void **)&d_invQ, sizeof(real) * M * M );
+    cudaMalloc( (void **)&d_testing, sizeof(real) * N * D );
+    cublasCheckErrors(cublasSetMatrix( M, D, sizeof(real), c_inputs, M, d_inputs, M ));
+    cublasCheckErrors(cublasSetMatrix( M, M, sizeof(real), c_invQ, M, d_invQ, M ));
+    cublasCheckErrors(cublasSetMatrix( N, D, sizeof(real), c_testing, N, d_testing, N));
+    
+
+        
+    /*********************************
```

Comment:
Make each of these chunks into a separate function. Always use functions to break up large routines rather than commenting section breaks.

This allows you to see inputs and outputs for sections clearly.

You can then use the function name to explain what the section does, rather than a comment.

A comment is good, but readable code is always better.